tools/buildsys/src/gomod.rs

```diff
@@ -0,0 +1,151 @@
+/*!
+Packages using the Go programming language may have upstream tar archives that
+include only the source code of the project, but not the source code of any
+dependencies. The Go programming language promotes the use of "modules" for
+dependencies and projects adopting modules will provide go.mod and go.sum
+files.
+
+This module extends the functionality of `packages.metadata.build-package.external-files`
+and provides the ability to retrieve and validate dependencies
+declared using Go modules given a tar archive containing a go.mod and go.sum.
+
+The location where dependencies are retrieved from are controlled by the
+standard environment variables employed by the Go tool: GOPROXY, GOSUMDB, and
+GOPRIVATE and are automatically retrieved from the host environment
+when docker-go is envoked.
+
+ */
+
+pub(crate) mod error;
+use error::Result;
+
+use super::manifest;
+use duct::cmd;
+use snafu::{ensure, OptionExt, ResultExt};
+use std::env;
+use std::path::{Path, PathBuf};
+use std::process::Output;
+
+pub(crate) struct GoMod;
+
+impl GoMod {
+    pub(crate) fn vendor(
+        root_dir: &Path,
+        package_dir: &Path,
+        gomod: &manifest::ExternalFile,
+    ) -> Result<()> {
+        let url_file_name = extract_file_name(&gomod.url)?;
+        let path_arg = &gomod.path.as_ref().unwrap_or(&url_file_name);
+        ensure!(path_arg.components().count() == 1, error::InputFileSnafu);
+
+        let path = package_dir.join(path_arg);
+        ensure!(path.is_file(), error::InputFileBadSnafu { path });
+
+        // If a module directory was not provided, set as an empty path.
+        // By default, without a provided module directory, tar will be passed
+        // the first directory found in the archieve as the top level Go module
+        let default_empty_path = PathBuf::from("");
+        let mod_dir = gomod.bundle_path.as_ref().unwrap_or(&default_empty_path);
```

Comment:
I had tried this originally but got hung up on the typing:

```
error[E0308]: mismatched types
  --> buildsys/src/gomod.rs:48:76
   |
48 |         let mod_dir = external_file.bundle_path.as_ref().unwrap_or_else(|| "".to_string()); // &default_empty_path);
   |                                                                            ^^^^^^^^^^^^^^ expected `&PathBuf`, found struct `std::string::String`
```

My understanding is that since `external_file_path` is a `PathBuf` and we are attempting to unwrap it, we would also need to return a pathbuf if it fails, not just an empty string? That's what lead me down the path of declaring an empty pathbuf and using it's borrowed value as the default if none was provided.